### PR TITLE
Skip zero-quota splits in get_dataset_samples

### PR DESCRIPTION
**Pull request by kevalmorabia97**
### What does this PR do?

Type of change: Bug fix

`get_dataset_samples` divides `num_samples` evenly across splits and dumps the remainder on the last one. When `num_samples` is smaller than the split count, most splits get a quota of `0` — but the old code still opened them:

```python
dataset_splits = [_load_split(s) for s in splits]   # all splits, eagerly
for dataset, n in zip(dataset_splits, num_per_split):
    for i, sample in enumerate(dataset):            # starts the iterator even when n == 0
        if i >= n:
            break
```

Merely starting a streamed split's iterator makes `datasets` fetch its first record batch, so a zero-quota split costs a real download and yields nothing.

For the `nemotron-post-training-v3` combo (14 samples over 7 datasets → 2 each) this is **24 split-opens that contribute zero samples**, including 4 splits of `nvidia/Nemotron-Math-v2`, whose parquet shards are ~12.6 GB each. Streaming those means reading a footer at byte ~12.6 GB and then pulling multi-MB row groups, all discarded.

This PR loads splits lazily and skips zero-quota ones. **Sample output is unchanged** — the remainder still lands on the last split, so the same rows are returned.

### Testing

`tests/gpu/torch/utils/test_dataset_utils.py::test_get_dataset_dataloader_nemotron_v3_chat_template` has been timing out on the 120 s `tests/gpu` cap in CI since 2026-07-31 — on `main`'s nightly and on every PR that runs the GPU job ([example](https://github.com/NVIDIA/Model-Optimizer/actions/runs/30657911036/job/91249036652)). The dumped stack shows it blocked in a socket read inside `hf_file_system._fetch_range` while streaming `Nemotron-Math-v2`.

Reproduced locally against the real gated datasets and measured:

| | runtime | result |
|---|---|---|
| before | 120.00 s | **FAIL** — same `Timeout (>120.0s)` as CI |
| after | 25 s | **PASS** |

- `tests/gpu/torch/utils/test_dataset_utils.py` — 7 passed
- `tests/unit/torch/utils/test_dataset_utils.py` — 49 passed

Note the timeout is not a recent code regression: the test ran 50–72 s for weeks, then stepped to >120 s on 2026-07-31 with no ModelOpt commit, dependency change (`datasets` 4.8.5 / `huggingface_hub` 1.14.0 / `hf-xet` 1.5.0 / `pyarrow` 24.0.0 identical across the boundary), or dataset-repo commit to explain it — Hub-side latency on Xet-backed range reads is the likely trigger. The wasted downloads were what left the test with no margin to absorb it; it now has ~4× headroom.

Also rejects negative `num_samples` at the interface boundary ([CodeRabbit](https://github.com/NVIDIA/Model-Optimizer/pull/2043#discussion_r3695434265)) — floor division turned `num_samples=-1` over 3 splits into quotas `[-1, -1, 1]`, silently returning one sample. Pre-existing on `main`, fixed here since it is adjacent to the quota logic.

### Before your PR is "*Ready for review*"

- Is this change backward compatible?: ✅
- If you copied code from any other sources or added a new PIP dependency, did you follow guidance in `CONTRIBUTING.md`: N/A
- Did you write any new necessary tests?: ✅ — `test_zero_quota_split_is_not_loaded` in `tests/unit/torch/utils/test_dataset_utils.py` (offline, uses the toy local dataset fixture); the existing `tests/gpu/torch/utils/test_dataset_utils.py` is the test this unblocks
- Did you update [Changelog](https://github.com/NVIDIA/Model-Optimizer/blob/main/CHANGELOG.rst)?: N/A — internal perf fix, no API or behavior change
- Did you get Claude approval on this PR?: ❌ — not yet run

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Enhancements**
  * Dataset loading now supports an optional split name.
  * Sample quotas are calculated independently for each split.
  * Negative sample counts are rejected with clear validation.
* **Performance Improvements**
  * Dataset splits load only when needed during sample iteration.
  * Splits with no assigned samples are skipped to avoid unnecessary initialization.

### CODE REVIEW

**Review by cjluo-nv — COMMENTED** [review]
> _Bot review (gpt-5.6-sol) — DM the bot to share feedback._

The lazy loading and `n <= 0` guard correctly preserve the existing quota/output ordering while avoiding all zero-quota split opens. The type widening for the default `None` split is also appropriate. Please add a small mocked unit regression test with fewer samples than splits that asserts `load_dataset` is called only for the positive-quota split(s); the existing live GPU test demonstrates the performance improvement but does not deterministically protect the behavior and depends on Hub latency.

File: modelopt/torch/utils/dataset_utils.py [hunk 1]
```diff
@@ -628,14 +628,15 @@ def _load_split(s: str):
...
         num_per_split[-1] += num_samples - sum(num_per_split)
 
         samples: list[str] = []
-        for dataset, n in zip(dataset_splits, num_per_split):
-            for i, sample in enumerate(dataset):
+        for split_name, n in zip(splits, num_per_split):
+            # Skip zero-quota splits: starting a streamed split fetches its first record batch.
+            if n <= 0:
+                continue
+            for i, sample in enumerate(_load_split(split_name)):
```
[the remark this answers is not in the document]
Comment (kevalmorabia97) [reply modelopt/torch/utils/dataset_utils.py:641]: Good catch — taken as suggested in 3747a435fa.

`num_samples` is now rejected at the interface boundary alongside the existing `DATASET_COMBOS` / `split` checks, and the quota skip is `n == 0` so a negative quota can never be silently swallowed:

```python
if num_samples < 0:
    raise ValueError(f"``num_samples`` must be non-negative, got {num_samples}.")
```

Worth noting this was pre-existing rather than introduced here: on `main`, `num_samples=-1` over 3 splits also returned one sample, since `i >= -1` breaks immediately on the first two splits. The new guard fixes it either way.

One deliberate divergence from the module's existing convention: the sibling helper `get_jsonl_text_samples` returns `[]` for `num_samples <= 0` rather than raising. I went with raising here per CONTRIBUTING.md's "validate external input once at the interface boundary", since a negative request is a caller bug worth surfacing on the public entry point. `num_samples=0` still returns `[]` as before.

**Review by cjluo-nv — APPROVED** [review]
> _Bot review (gpt-5.6-sol) — DM the bot to share feedback._

Re-review complete. The prior critical test-coverage concern is addressed by the new offline unit regression test: with one sample across two splits, it spies on `datasets.load_dataset` and verifies that only the positive-quota `test` split is opened. The lazy loading and zero-quota skip preserve quota/output ordering, and the negative `num_samples` validation resolves the separate negative-quota edge case. No new correctness or licensing concerns found.
